ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -109,81 +117,154 @@ codeLensProvider :: PluginMethodHandler IdeState Method_TextDocumentCodeLens
+
+-- Since the lenses are created with diagnostics, and since the globalTypeSig
+-- rule can't be changed as it is also used by the hls-refactor plugin, we can't
+-- rely on actions. Because we can't rely on actions it doesn't make sense to
+-- recompute the edit upon command. Hence the command here just takes a edit
+-- and applies it.
 commandHandler :: CommandFunction IdeState WorkspaceEdit
 commandHandler _ideState wedit = do
   _ <- lift $ LSP.sendRequest SMethod_WorkspaceApplyEdit (ApplyWorkspaceEditParams Nothing wedit) (\_ -> pure ())
   pure $ InR Null
 
 --------------------------------------------------------------------------------
-
-suggestSignature :: Bool -> Maybe HscEnv -> Maybe GlobalBindingTypeSigsResult -> Maybe TcModuleResult -> Maybe Bindings -> Diagnostic -> [(T.Text, [TextEdit])]
+-- To give one an idea about how creative hls-refactor plugin is, the end type
+-- here can be changed within certain parameters, and even though it is used by
+-- the hls-refactor-plugin, the hls-refactor-plugin itself won't need adaptions
+suggestSignature :: Bool -> Maybe HscEnv -> Maybe GlobalBindingTypeSigsResult -> Maybe TcModuleResult -> Maybe Bindings -> Diagnostic -> [(T.Text, TextEdit)]
 suggestSignature isQuickFix env mGblSigs mTmr mBindings diag =
-  suggestGlobalSignature isQuickFix mGblSigs diag <> suggestLocalSignature isQuickFix env mTmr mBindings diag
-
-suggestGlobalSignature :: Bool -> Maybe GlobalBindingTypeSigsResult -> Diagnostic -> [(T.Text, [TextEdit])]
+  maybeToList (suggestGlobalSignature isQuickFix mGblSigs diag) <> maybeToList (suggestLocalSignature isQuickFix env mTmr mBindings diag)
+
+-- Both the suggestGlobalSignature and suggestLocalSignature functions have been
+-- broken up. The main functions works with a diagnostic, which then calls the
+-- secondary function with whatever pieces of the diagnostic it needs. This
+-- allows the resolve function, which no longer has the Diagnostic, to still
+-- call the secondary functions.
+suggestGlobalSignature :: Bool -> Maybe GlobalBindingTypeSigsResult -> Diagnostic -> Maybe (T.Text, TextEdit)
 suggestGlobalSignature isQuickFix mGblSigs Diagnostic{_message, _range}
-  | _message
-      =~ ("(Top-level binding|Pattern synonym) with no type signature" :: T.Text)
-    , Just (GlobalBindingTypeSigsResult sigs) <- mGblSigs
-    , Just sig <- find (\x -> sameThing (gbSrcSpan x) _range) sigs
+  | _message =~ ("(Top-level binding|Pattern synonym) with no type signature" :: T.Text) =
+    suggestGlobalSignature' isQuickFix mGblSigs Nothing _range
+  | otherwise = Nothing
+
+-- In addition, for suggestGlobalSignature, we added the option of having a
+-- PositionMapping. In this case if there is no PositionMapping provided, it will
+-- ignore it. However if a PositionMapping is supplied, it will assume that the
+-- range provided is already converted with the PositionMapping, and will attempt
+-- to convert it back before attempting to find the signature from the rule.
```

Comment:
This feels a little funny - if the Range needs to be predictably converted in this way, why can't it be the caller's responsibility?